File: DotNetCorePracticalView/CS.DotNetCore.LoadTest/src/CS.DotNetCore.LoadTest.WebApp/Data/Pgsql/IdentityPgsqlDAO .cs
namespace CS.DotNetCore.LoadTest.WebApp.Data.Pgsql
{
    using System.Threading.Tasks;
    using Business;
    using Config;
    using Npgsql;
    using Newtonsoft.Json;
    using System.Data;
    using NpgsqlTypes;
    using Schema;
    using Util.Data;

    internal class IdentityPgsqlDAO : IIdentityAsyncDAO
    {
        private static readonly string DeleteSqlCommand = string.Format
            ("delete from \"{0}\";", IdentitySchema.Table);

        private static readonly string InsertSqlCommand = string.Format
            ("insert into \"{0}\" (\"{1}\",\"{2}\") values (:{1}, :{2});",
            IdentitySchema.Table,
            IdentitySchema.IdentityNameColumn,
            IdentitySchema.DataColumn);

        private IDatabaseConfig _config;

        public IdentityPgsqlDAO(IDatabaseConfig config)
        {
            _config = config;
        }

        private static NpgsqlParameter[] CreateInsertParameters(Identity identity)
        {
            return new NpgsqlParameter[2]
            {
                new NpgsqlParameter(IdentitySchema.IdentityNameColumn, NpgsqlDbType.Varchar)
                {
                    Direction = ParameterDirection.Input,
                    NpgsqlValue = identity.IdentityName,
                    ParameterName = IdentitySchema.IdentityNameColumn
                },

                new NpgsqlParameter(IdentitySchema.DataColumn, NpgsqlDbType.Json)
                {
                    Direction = ParameterDirection.Input,
                    IsNullable = false,
                    NpgsqlValue = JsonConvert.SerializeObject(identity),
                }
            };
        }

        public async Task InsertAsync(Identity identity)
        {
            using (var conn = new NpgsqlConnection(_config.String))
            {
                await conn.OpenAsync().ConfigureAwait(false);
                await DataHelper.ExecuteNonQueryAsync(conn, InsertSqlCommand, CreateInsertParameters(identity)).ConfigureAwait(false);
            }
        }

        public async Task<int> DeleteAllAsync()
        {
            using (var conn = new NpgsqlConnection(_config.String))
            {
                await conn.OpenAsync().ConfigureAwait(false);
                return await DataHelper.ExecuteNonQueryAsync(conn, DeleteSqlCommand).ConfigureAwait(false);
            }
        }
    }
}

File: DotNetCorePracticalView/CS.NodeJS.LoadTest.WebApp/webapp/data/pgsql/identityPgsqlDAO.js
let schema = require('../loadTestSchema.js')

module.exports = {

  _deleteAllSqlCommand: 'delete from "' + schema.identity.table + '";',

  _insertSqlCommand: 'insert into "' + schema.identity.table + '" ("' + schema.identity.identityNameColumn +
  '","' + schema.identity.dataColumn + '") values ($1, $2);',

  insertAsync: (identity, connPool) => {

    let queryConfig = {
      text: module.exports._insertSqlCommand,
      values: [identity.identityName, identity]
    };

    return new Promise((res, rej) => {

      connPool.connect()
        .then(cl => {

          let insert = cl.query(queryConfig, (err, result) => {

            cl.release();

            if (err) {
              rej(err);
            }
            else {
              res();
            }
          });
        })
        .catch(rej);
    });
  },

  deleteAllAsync: (connPool) => {

    let queryConfig = {
      text: module.exports._deleteAllSqlCommand,
      values: []
    };

    return new Promise((resolve, reject) => {

      connPool.connect()
        .then(cl => {

          let deleteCmd = cl.query(queryConfig, (err, result) => {

            cl.release();

            if (err) {
              reject(err);
            }
            else {
              resolve();
            }
          });
        })
        .catch(reject)
    });
  }
}